src/builtins.ts

```diff
@@ -6565,6 +6686,78 @@ function builtin_f64_div(ctx: BuiltinContext): ExpressionRef {
 }
 builtins.set(BuiltinNames.f64_div, builtin_f64_div);
 
+// i32.eq -> eq<i32>
+function builtin_i32_eq(ctx: BuiltinContext): ExpressionRef {
+  checkTypeAbsent(ctx);
+  ctx.typeArguments = [ Type.i32 ];
+  ctx.contextualType = Type.i32;
+  return builtin_eq(ctx);
+}
+builtins.set(BuiltinNames.i32_eq, builtin_i32_eq);
+
+// i64.eq -> eq<i64>
+function builtin_i64_eq(ctx: BuiltinContext): ExpressionRef {
+  checkTypeAbsent(ctx);
+  ctx.typeArguments = [ Type.i64 ];
+  ctx.contextualType = Type.i32;
+  return builtin_eq(ctx);
+}
+builtins.set(BuiltinNames.i64_eq, builtin_i64_eq);
+
+// f32.eq -> eq<f32>
+function builtin_f32_eq(ctx: BuiltinContext): ExpressionRef {
+  checkTypeAbsent(ctx);
+  ctx.typeArguments = [ Type.f32 ];
+  ctx.contextualType = Type.i32;
+  return builtin_eq(ctx);
+}
+builtins.set(BuiltinNames.f32_eq, builtin_f32_eq);
+
+// f64.eq -> eq<f64>
+function builtin_f64_eq(ctx: BuiltinContext): ExpressionRef {
+  checkTypeAbsent(ctx);
+  ctx.typeArguments = [ Type.f64 ];
+  ctx.contextualType = Type.i32;
+  return builtin_eq(ctx);
+}
+builtins.set(BuiltinNames.f64_eq, builtin_f64_eq);
+
+// i32.ne -> ne<i32>
+function builtin_i32_ne(ctx: BuiltinContext): ExpressionRef {
+  checkTypeAbsent(ctx);
+  ctx.typeArguments = [ Type.i32 ];
+  ctx.contextualType = Type.i32;
+  return builtin_ne(ctx);
+}
+builtins.set(BuiltinNames.i32_ne, builtin_i32_ne);
+
+// i64.ne -> ne<i64>
+function builtin_i64_ne(ctx: BuiltinContext): ExpressionRef {
+  checkTypeAbsent(ctx);
+  ctx.typeArguments = [ Type.i64 ];
+  ctx.contextualType = Type.i64;
```

Comment:
The result of `i64.ne` is
```suggestion
  ctx.contextualType = Type.i32;
```